File: programming-assignments/programming-assignment-2/code/pa2.cpp
#include <vector>
#include <algorithm>
#include <stdlib.h> // for testing
#include <time.h> // for random seed
#include <algorithm> // also for testing
#include <iostream>

class Point {
public:
	int x, y;
	int distance_from_origin;

	Point(int x1, int y1)
	{
		x = x1;
		y = y1;
		distance_from_origin = x * x + y * y;
	}
	bool operator<(const Point& p)
	{
		return distance_from_origin < p.distance_from_origin;
	}

	bool operator!=(const Point& p)
	{
		return x != p.x or y != p.y;
	}
};

using namespace std;

Point median(vector<Point> points)
{
	/* median of vector of points by distance to origin
	 * assume O(1) length
     */
	int n = points.size();
	std::sort(points.begin(), points.end());
	return points[n / 2];
}

vector<Point> medians(vector<Point>& points)
{
	/*Returns the set of medians of disjoint subarrays of length 5*/
	vector<Point> ret;
	for (auto i = 0; i < points.size(); i += 5) {
		auto start = points.begin() + i;
		auto end = (start + 5 < points.end()) ? start + 5 : points.end();
		vector<Point> sub_vec(start, end);
		auto median_of_5 = median(sub_vec);
		ret.push_back(median_of_5);
	}
	return ret;
}

Point median_of_medians(vector<Point> points, int k)
{
	if (points.size() == 1) {
		return points[0];
	} else {
		auto meds = medians(points);
		auto pivot = median_of_medians(meds, meds.size() / 2);
		vector<Point> L, P, R;
		for (auto p : points) {
			if (p < pivot) {
				L.push_back(p);
			} else if (pivot < p) {
				R.push_back(p);
			} else {
				P.push_back(p);
			}
		}
		if (L.size() == k - 1) { // pivot is the kth smallest
			return pivot;
		} else if (L.size() > k - 1) { // kth smallest is in left
			return median_of_medians(L, k);
		} else { // kth smallest is in right
			return median_of_medians(R, k - L.size() - 1);
		}
	}
}

int main()
{
	/* Generate random points and a random k and test correctness
     * by comparing output against the built-in sorting algorithm
     */
	srand(time(NULL));
	for (int _ = 0; _ < 100; _++) {
		int n = rand() % 100 + 100;
		int k = rand() % n + 1;
		vector<Point> points;
		for (int __ = 0; __ < n; __++) {
			points.push_back(Point(rand(), rand()));
		}
		Point select_answer = median_of_medians(points, k);
		sort(points.begin(), points.end());
		Point sort_answer = points[k - 1];
		if (sort_answer != select_answer) { // found a possible error
			for (int i = 0; i < n; i++) {
				cout << "(" << points[i].x << ", " << points[i].y << ") ";
			}
			cout << endl;
			cout << "(" << sort_answer.x << ", " << sort_answer.y << ") ";
			cout << "(" << select_answer.x << ", " << select_answer.y << ") ";
			return -1;
		}
	}
	cout << "Success" << endl;
	return 0;
}

File: programming-assignments/programming-assignment-2/code/pa2.py
#!/usr/bin/env python3

from random import random, randint  # for testing


class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.tup = (self.x, self.y)
        self.square_distance_from_origin = self.x ** 2 + self.y ** 2

    def __lt__(self, q):
        return self.square_distance_from_origin < q.square_distance_from_origin


def medians(A, size=5):
    """Returns the set of medians of disjoint subarrays of length size """
    partition = [sorted(A[i : i + size]) for i in range(0, len(A), size)]
    return [subarr[len(subarr) // 2] for subarr in partition]


def median_of_medians(points, k, size=5):
    if len(points) == 1:
        return points[0]
    meds = medians(points, size)
    pivot = median_of_medians(medians(points), len(meds) // 2)
    left, right = [], []
    for p in points:
        if p < pivot:
            left.append(p)
        elif pivot < p:
            right.append(p)
    if len(left) == k - 1:  # pivot is the kth smallest
        return pivot
    elif len(left) > k - 1:  # kth smallest is in left
        return median_of_medians(left, k)
    else:  # kth smallest is in right
        return median_of_medians(right, k - len(left) - 1)


def main():
    """Generate random points and a random k and test correctness
    by comparing output against the built-in sorting algorithm
    """
    for _ in range(100):
        n = randint(100, 200)
        k = randint(1, n)
        points = [Point(random(), random()) for _ in range(n)]
        sort_answer = sorted(points)[k - 1]
        select_answer = median_of_medians(points, k)
        if sort_answer != select_answer:  # found a possible error
            print(points)
            print(sort_answer)
            print(select_answer)
            return
    print("Success!")


if __name__ == "__main__":
    main()
